### PR TITLE
Fix issues with ROCm memory registration

### PR DESCRIPTION
**Pull request by souravzzz**
## What
Fix issues with ROCm memory registration

## How ?
1. Fix incorrect logic for identifying device memory
2. Don't try to lock device memory for registration

### CODE REVIEW

File: src/uct/rocm/copy/rocm_copy_md.c [hunk 1]
```diff
@@ -92,21 +92,53 @@ static ucs_status_t uct_rocm_copy_rkey_release(uct_component_t *component,
...
         return UCS_OK;
     }
 
-    status = hsa_amd_memory_lock(address, length, NULL, 0, &dev_addr);
-    if ((status != HSA_STATUS_SUCCESS) || (dev_addr == NULL)) {
-        return UCS_ERR_IO_ERROR;
+    err = uct_rocm_base_detect_memory_type(uct_md, address, length, &mem_type);
+    if (err != UCS_OK) {
+        ucs_error("failed to detect memory type for addr %p len %zu", address, length);
+        ucs_free(mem_hndl);
```
Comment (yosefe) [inline src/uct/rocm/copy/rocm_copy_md.c:126]: seems like double free

File: src/uct/rocm/ipc/rocm_ipc_md.c [hunk 2]
```diff
@@ -54,14 +54,15 @@ static ucs_status_t uct_rocm_ipc_mkey_pack(uct_md_h md, uct_mem_h memh,
...
     hsa_status_t status;
     hsa_agent_t agent;
-    void *base_ptr;
-    size_t size;
 
     status = uct_rocm_base_get_ptr_info(address, length, &base_ptr, &size, &agent);
-    if (status != HSA_STATUS_SUCCESS) {
-        ucs_error("pack none ROCM ptr %p/%lx", address, length);
+    if ((status != HSA_STATUS_SUCCESS) || (size < length)) {
+        ucs_error("Failed to get base ptr for %p/%lx, ROCm returned %p/%lx",
```
Comment (yosefe) [inline src/uct/rocm/ipc/rocm_ipc_md.c:64]: smallcase

File: src/uct/rocm/ipc/rocm_ipc_md.c [hunk 3]
```diff
@@ -54,14 +54,15 @@ static ucs_status_t uct_rocm_ipc_mkey_pack(uct_md_h md, uct_mem_h memh,
 static hsa_status_t uct_rocm_ipc_pack_key(void *address, size_t length,
                                           uct_rocm_ipc_key_t *key)
 {
+    void *base_ptr = NULL;
+    size_t size = 0;
```
Comment (yosefe) [inline src/uct/rocm/ipc/rocm_ipc_md.c:58]: align on = on same column

File: src/uct/rocm/base/rocm_base.c [hunk 4]
```diff
@@ -185,18 +185,11 @@ ucs_status_t uct_rocm_base_detect_memory_type(uct_md_h md, const void *addr,
...
-            (dev_type == HSA_DEVICE_TYPE_GPU)) {
-            *mem_type_p = UCS_MEMORY_TYPE_ROCM;
-            return UCS_OK;
-        }
+        (info.type != HSA_EXT_POINTER_TYPE_UNKNOWN)) {
+        *mem_type_p = UCS_MEMORY_TYPE_ROCM;
     }
 
-    return UCS_ERR_INVALID_ADDR;
+    return UCS_OK;
```
Comment (yosefe) [inline src/uct/rocm/base/rocm_base.c:192]: so now we can return OK, but `*mem_type_p` would not be initialized?

Comment (souravzzz) [reply src/uct/rocm/base/rocm_base.c:192]: *mem_type_p is always initialized to UCS_MEMORY_TYPE_HOST in line 180

File: src/uct/rocm/copy/rocm_copy_md.c (hunk 1 above)
Comment (souravzzz) [reply src/uct/rocm/copy/rocm_copy_md.c]: Fixed

File: src/uct/rocm/ipc/rocm_ipc_md.c (hunk 2 above)
Comment (souravzzz) [reply src/uct/rocm/ipc/rocm_ipc_md.c]: done

File: src/uct/rocm/ipc/rocm_ipc_md.c (hunk 3 above)
Comment (souravzzz) [reply src/uct/rocm/ipc/rocm_ipc_md.c]: done

**Review by yosefe — APPROVED** [review]
pls squash

**Comment by souravzzz** [comment]
@yosefe Squashed the memory registration related changes.
Had to include another commit 61156bb57170f6970ffd4ab4900af2a23638e362 to fix issues introduced by 98a614188c026b7682f914ef62864a24fb44b464 Can you please review?

**Review by yosefe — APPROVED** [review]

**Pull request merged**